### PR TITLE
[charts] Introduce keyboard navigation

### PR DESCRIPTION
**Pull request by alexfauquette**
Introduce keyboard navigation

The data format is fairly similar to the highlight plugin

Rendering aspect

- The line series get a dedicated component to show the highlighted point since marks are not always rendered.
- All the other series implemented use the `outlined: 'auto'` property to show the focus position. And they rely on `data-focused=true` for that.

Focus management

- Zoom is not taken into account. Should we loop only on visible points, or move the zoom to have the focused element inside the drawing area?
- The vocal assistant strategy I'm planning to use is to use `aria-activedescendent` to keep the focus on the SVG and target specific items like I'm already doing it with `data-focused=true`. I will probably add an `id` attribute only on the element of interest.
- Will need an option to disable this keyboard navigation, such that user can do whatever they want. For eample the npm sparkline reproduction will be broken because of this feature (the focus can not be handled at the same time by the SVG and a parent div)

### CODE REVIEW

File: packages/x-charts/src/BarChart/BarChart.tsx [hunk 1]
```diff
@@ -140,7 +140,7 @@ const BarChart = React.forwardRef(function BarChart(
       <ChartsWrapper {...chartsWrapperProps}>
         {props.showToolbar && Toolbar ? <Toolbar {...props.slotProps?.toolbar} /> : null}
         {!props.hideLegend && <ChartsLegend {...legendProps} />}
-        <ChartsSurface {...chartsSurfaceProps}>
+        <ChartsSurface {...chartsSurfaceProps} tabIndex={0}>
```
Comment (JCQuintas) [inline packages/x-charts/src/BarChart/BarChart.tsx]: Is this `tabIndex` a mistake? 🤔

File: packages/x-charts/src/ChartsSurface/ChartsSurface.tsx [hunk 2]
```diff
@@ -78,6 +83,7 @@ const ChartsSurface = React.forwardRef<SVGSVGElement, ChartsSurfaceProps>(functi
       className={className}
       {...other}
       ref={handleRef}
+      tabIndex={0}
```
Comment (JCQuintas) [inline packages/x-charts/src/ChartsSurface/ChartsSurface.tsx]: Should we move this up so users can disable it? 🤔

File: packages/x-charts/src/hooks/useItemFocused.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * A hook to check if an item has the focus.
+ *
+ * If you need to process multiple points, use the `useItemFocusedGetter` hook instead.
+ *
+ * @param {FocusedItemData} item is the item to check
+ * @returns {boolean} the focus state
+ */
+export function useItemFocused(item: UseItemFocusedParams): boolean {
```
Comment (JCQuintas) [inline packages/x-charts/src/hooks/useItemFocused.ts]: Should we use `useIsItemFocused` here?

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.selectors.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+export const selectorChartsFocusedSeriesId = createSelector(
+  [selectKeyboardNavigation],
+  (item) => item?.seriesId,
+);
+
+export const selectorChartsFocusedDataIndex = createSelector(
+  [selectKeyboardNavigation],
+  (item) => item?.dataIndex,
+);
```
Comment (JCQuintas) [inline packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.selectors.ts]: Should we already account for non-regular identifiers here? 😉 

This feels a bit static if we consider the identifier is configurable. Maybe we can have a more generic selector with a proper memo function? 🤔 

Maybe we should have a `compareIdentifier` in the series config 💡

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        return {
+          type: typesAvailable[typeIndex],
+          seriesId: seriesOfType.seriesOrder[seriesIndex],
+        };
+      }
+    }
+  }
+
+  return null;
+}
```
Comment (JCQuintas) [inline packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts]: Should we move these big functions to another file?

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+} | null {
+  const startingTypeIndex =
+    type !== undefined && series[type] ? Object.keys(series).indexOf(type) : 0;
+  const currentSeriesIndex =
+    type !== undefined && seriesId !== undefined && series[type] && series[type].series[seriesId]
+      ? series[type].seriesOrder.indexOf(seriesId)
+      : -1;
+  const typesAvailable = Object.keys(series) as (keyof typeof series)[];
+
+  for (let typeGap = 0; typeGap < typesAvailable.length; typeGap += 1) {
```
Comment (JCQuintas) [inline packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts]: Would be good to have a brief description of what these loops do

```
// Check if other series of the same type has data

// Check if series of other type has data
```

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+          }
+
+          break;
+        }
+        case 'ArrowUp': {
+          const updatedStore = focusNextSeries();
+
+          if (updatedStore) {
+            event.preventDefault();
+          }
```
Comment (JCQuintas) [inline packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts:227]: Should we comment that these are to prevent scroll?

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+          break;
+        }
+        default:
+          break;
+      }
+    }
+
+    element.addEventListener('keydown', keyBoardHandler);
+    element.addEventListener('focus', focusNextItem);
```
Comment (JCQuintas) [inline packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts]: I don't think focus should do anything 🤔 

With focus whenever a user clicks anywhere in the chart, the first item will be selected, which I think is quite odd.

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.types.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  type: ChartSeriesType;
+  /**
+   * The series id of the funnel.
+   */
+  seriesId: SeriesId;
+  /**
+   * The index of the data point in the series.
+   */
+  dataIndex: number;
+};
```
Comment (JCQuintas) [inline packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.types.ts]: Should we use the one defined in the series config? Also this is not for the funnel 😆

File: packages/x-charts/src/LineChart/FocusedMark.tsx [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+  return (
+    <rect
+      fill="none"
+      stroke={(theme.vars ?? theme).palette.text.primary}
+      strokeWidth={2}
+      x={xAxis[xAxisId].scale(xAxis[xAxisId].data![focusedItem.dataIndex])! - RADIUS}
+      y={yAxis[yAxisId].scale(series.stackedData[focusedItem.dataIndex][1])! - RADIUS}
+      width={2 * RADIUS}
+      height={2 * RADIUS}
```
Comment (JCQuintas) [inline packages/x-charts/src/LineChart/FocusedMark.tsx:33]: I assume we can't get the proper size of the original mark? Using a static radius is a bit weird

File: packages/x-charts/src/BarChart/BarChart.tsx (hunk 1 above)
Comment (alexfauquette) [reply packages/x-charts/src/BarChart/BarChart.tsx]: Yes it's an intial experiment I forgot to remove

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.selectors.ts (hunk 4 above)
Comment (alexfauquette) [reply packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.selectors.ts]: > This feels a bit static if we consider the identifier is configurable

I assume you have scatter charts in mind ;)

The identifier will not be the biggest issue. it's the keyboard actions. For those who does not enter in the `(seriesId, dataIndex)` identifier framework, it might be easier to create a custom plugin for their keyboard navigation.

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts (hunk 8 above)
Comment (alexfauquette) [reply packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts]: It's inspired by ARIA guidelines from the data grid. When the element get focused, it set the focus to it's first sub element.

I could have a rule saying that if charts is focused, but has no internal element focus, I dislay the outlined on the entire SVG. But when user click on th charts I have to show it's getting the focus

File: packages/x-charts/src/LineChart/FocusedMark.tsx (hunk 10 above)
Comment (alexfauquette) [reply packages/x-charts/src/LineChart/FocusedMark.tsx:33]: Line chart don't have a clean way to set size. The only option for users is to create their own elements. But in that case we don't have info about their size.

https://github.com/mui/mui-x/blob/master/packages/x-charts/src/LineChart/CircleMarkElement.tsx/#L85

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts (hunk 6 above)
Comment (alexfauquette) [reply packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts]: Added them, and noticed bad edge cases, I fixed+test

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.types.ts (hunk 9 above)
Comment (alexfauquette) [reply packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.types.ts:20]: What do you mean by "This is not for the funnel" ?

```ts
/**
 * An object that allows to identify a funnel item.
 * Used for item interaction
 */
export type FunnelItemIdentifier = {
  type: 'funnel';
  /**
   * The series id of the funnel.
   */
  seriesId: SeriesId;
  /**
   * The index of the data point in the series.
   */
  dataIndex: number;
};
```

Comment (alexfauquette) [reply packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.types.ts]: It took some time to spot why you talked about the funnel

File: docs/data/charts/accessibility/accessibility.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+Both `Desktop` and `Mobile` Date and Time Pickers are using `role="dialog"` to display their interactive view parts and thus they should follow [Modal accessibility guidelines](/material-ui/react-modal/#accessibility).
+This behavior is automated as much as possible, ensuring that the Date and Time Pickers are accessible in most cases.
+A correct `aria-labelledby` value is assigned to the dialog component based on the following rules:
+
+- Use `toolbar` id if the toolbar is visible;
+- Use the id of the input label if the toolbar is hidden;
+
+:::info
+Make sure to provide an `aria-labelledby` prop to `popper` and/or `mobilePaper` slots in case you are using Date and Time Pickers component with **hidden toolbar** and **without a label**.
+::: -->
```
Comment (JCQuintas) [inline docs/data/charts/accessibility/accessibility.md]: should we remove this? Or maybe just leave the titles in for these commented out sections

File: docs/data/charts/accessibility/accessibility.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+## Animation
+
+Most of the charts have an animation when rendering or when data update.
```
Comment (JCQuintas) [inline docs/data/charts/accessibility/accessibility.md]:
```suggestion
Some charts have animations when rendering or when data updates.
```

futureproofing 😅

File: docs/data/charts/accessibility/accessibility.md [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+For users with vestibular motion disorders those animations can be problematic.
+By default we skip animation based on the [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) media feature.
```
Comment (JCQuintas) [inline docs/data/charts/accessibility/accessibility.md]:
```suggestion
By default animations are toggled based on the [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) media feature.
```

File: docs/data/charts/accessibility/accessibility.md [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+Common conformance guidelines for accessibility include:
+
+- Globally accepted standard: [WCAG](https://www.w3.org/WAI/standards-guidelines/wcag/)
+- US:
+  - [ADA](https://www.ada.gov/) - US Department of Justice
+  - [Section 508](https://www.section508.gov/) - US federal agencies
+- Europe: [EAA](https://employment-social-affairs.ec.europa.eu/policies-and-activities/social-protection-social-inclusion/persons-disabilities/union-equality-strategy-rights-persons-disabilities-2021-2030/european-accessibility-act_en) (European Accessibility Act)
+
+WCAG 2.1 has three levels of conformance: A, AA, and AAA.
+Level AA exceeds the basic criteria for accessibility and is a common target for most organizations, so this is what we aim to support.
```
Comment (JCQuintas) [inline docs/data/charts/accessibility/accessibility.md]:
```suggestion
Level AA exceeds the basic criteria for accessibility and is a common target for most organizations, so this is what this library aims to support.
```

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.helpers.test.ts [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+        id: 'a',
+        ...lineSeries,
+      },
+    },
+    seriesOrder: ['a', 'b'],
+    stackingGroups: [],
+  },
+};
+
+describe('useChartKeyboardNavigation - helpers', () => {
```
Comment (JCQuintas) [inline packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.helpers.test.ts:77]: 👍

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts (hunk 8 above)
Comment (JCQuintas) [reply packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts]: Yeah that would be less disruptive I think. 🤔

File: docs/data/charts/accessibility/accessibility.md [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+<!--
+## Screen reader compatibility
+
+Date and Time Pickers use ARIA roles and robust focus management across the interactive elements to convey the necessary information to users, being optimized for use with assistive technologies.
+-->
+
+## Keyboard support
+
+Set `enableKeyboardNavigation` to `true` to enable the keyboard navigation on your charts.
```
Comment (JCQuintas) [inline docs/data/charts/accessibility/accessibility.md:59]: Should we have a global enable/disable toggle for this? 🤔

File: docs/data/charts/accessibility/accessibility.md [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+<p class="description">Learn how the Charts implement accessibility features and guidelines, including keyboard navigation that follows international standards.</p>
+
+:::info
+A common misconception about accessibility is to only consider blind people and the screen reader.
+But there are other disability to consider, like:
+
+- **Color blindness**, making it hard to distinguish different series, or low contrast elements.
+- **Motion disability**, making it hard to open the tooltip on a given item.
+- **Cognitive disability**, making it hard to focus your attention on some details.
+- **Vestibular dysfunction**, making you unconformable with animations.
```
Comment (bernardobelchior) [inline docs/data/charts/accessibility/accessibility.md]:
```suggestion
- **Vestibular dysfunction**, making you uncomfortable with animations.
```

File: docs/data/charts/accessibility/accessibility.md (hunk 16 above)
Comment (bernardobelchior) [inline docs/data/charts/accessibility/accessibility.md:59]: Should we add a demo?

File: packages/x-charts/src/BarChart/BarElement.tsx [hunk 18]
```diff
@@ -103,6 +110,7 @@ function BarElement(props: BarElementProps) {
       fill: color,
       skipAnimation,
       layout,
+      'data-focused': isFocused || undefined,
```
Comment (bernardobelchior) [inline packages/x-charts/src/BarChart/BarElement.tsx:113]: [Base UI](https://base-ui.com/react/components/menu) uses `data-highlighted`, but that might conflict with our current highlighted logic. 

Image

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+    });
+  });
+
+  React.useEffect(() => {
+    const element = svgRef.current;
+    if (!element || !params.enableKeyboardNavigation) {
+      return undefined;
+    }
+
+    function keyBoardHandler(event: KeyboardEvent) {
```
Comment (bernardobelchior) [inline packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts]: Keyboard is a single word

```suggestion
    function keyboardHandler(event: KeyboardEvent) {
```

**Review by bernardobelchior — COMMENTED** [review]
I'm having a hard time using this feature in the docs. I can't seem to focus the chart. Do you know how I can do that? Do I need to enable some specific feature?

**Comment by alexfauquette** [comment]
> Do I need to enable some specific feature?

Yes, you need to add `enableKeyboardNavigation` prop to the chart. It should be enable by default but this would be a breaking change

**Comment by bernardobelchior** [comment]
> > Do I need to enable some specific feature?
> 
> Yes, you need to add `enableKeyboardNavigation` prop to the chart. It should be enable by default but this would be a breaking change

I see. Should we add a demo, then? Otherwise people won't be able to easily test it from the docs

**Comment by bernardobelchior** [comment]
Seems to work very well 😄 

https://github.com/user-attachments/assets/69a62c7b-bd60-4332-8aed-26df9e5f3a13

Do we want to show the tooltip when a mark is focused and the user presses some key?

File: docs/data/charts/accessibility/accessibility.md (hunk 11 above)
Comment (alexfauquette) [inline docs/data/charts/accessibility/accessibility.md]:
```suggestion
```

File: docs/data/charts/accessibility/accessibility.md (hunk 16 above)
Comment (alexfauquette) [reply docs/data/charts/accessibility/accessibility.md:59]: Done :+1:

File: packages/x-charts/src/BarChart/BarElement.tsx (hunk 18 above)
Comment (alexfauquette) [reply packages/x-charts/src/BarChart/BarElement.tsx:113]: `data-highlighted` could clearly conflict. But maybe `data-active` to use similar wording as `aria-activedescendent`

File: docs/data/charts/accessibility/accessibility.md (hunk 16 above)
Comment (alexfauquette) [reply docs/data/charts/accessibility/accessibility.md:59]: I added a sentence about how to use theme provider to do so

File: packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts (hunk 8 above)
Comment (JCQuintas) [reply packages/x-charts/src/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.ts]: Did you do anything regarding this comment? I still see the blue around it when clicking on the chart

File: docs/data/charts/accessibility/accessibility.md [hunk 20]
```diff
@@ -0,0 +1,74 @@
+---
+productId: x-charts
+title: Charts - Accessibility
+packageName: '@mui/x-charts'
+---
+
+# Accessibility
+
+<p class="description">Learn how the Charts implement accessibility features and guidelines, including keyboard navigation that follows international standards.</p>
```
Comment (JCQuintas) [inline docs/data/charts/accessibility/accessibility.md:9]:
```suggestion
<p class="description">Learn how Charts implement accessibility features and guidelines, including keyboard navigation that follows international standards.</p>
```

File: docs/data/charts/accessibility/accessibility.md [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+  },
+}
+```
+
+{{"demo": "KeyboardNavigation.js"}}
+
+This feature is currently supported by line, bar, pie, scatter, and sparkline charts.
+
+This makes the SVG component focusable thanks to [`tabIndex`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Global_attributes/tabindex).
+When focused, the charts highlight a value item that can be modified with arrow navigation.
```
Comment (JCQuintas) [inline docs/data/charts/accessibility/accessibility.md]:
```suggestion
When focused, the chart highlights a value item that can be modified with arrow navigation.
```

File: packages/x-charts/src/ChartsSurface/ChartsSurface.tsx [hunk 22]
```diff
@@ -45,9 +45,16 @@ const ChartsSurfaceStyles = styled('svg', {
...
+  '&:focus': {
+    outline: 'none', // By default don't show focus on the SVG container
+  },
+  '&:focus-visible': {
+    // Show focus outline on the SVG container only when using keyboard navigation
+    outline: 'auto',
+    '&[data-has-focused-item=true]': {
+      // But not if the chart has a focused children item
+      outline: 'none',
+    },
```
Comment (alexfauquette) [inline packages/x-charts/src/ChartsSurface/ChartsSurface.tsx:57]: @JCQuintas I went further with the focus display.

By relying on the `:focus-visible` the outlined is not display when the element get focused by click, but is visible when focused by <kbd>Tab</kbd>

Comment (JCQuintas) [reply packages/x-charts/src/ChartsSurface/ChartsSurface.tsx:57]: Nice, interactions feels good now, thanks :D

**Review by JCQuintas — APPROVED** [review]
small docs suggestions, everything else looks good

File: docs/data/charts/accessibility/accessibility.md [hunk 23]
```diff
@@ -48,6 +48,13 @@ Date and Time Pickers use ARIA roles and robust focus management across the inte
 
 ## Keyboard support
 
+:::warning
+This feature is under development.
+The way keyboard interaction is visualized will evolve.
+
+For example the element highlight, or tooltip will be impacted by the feature.
```
Comment (alexfauquette) [inline docs/data/charts/accessibility/accessibility.md:55]:
```suggestion
For example the element highlight, or tooltip will be impacted by the feature.
Those modifications will not be considered as breaking changes and can be added during minor or patch version
```
